### PR TITLE
debouncing of daily production

### PR DESCRIPTION
**Pull request by patman15**
Hey @mawoka-myblock,

here is my proposal (attempt #4) to fix the restart of the daily production (fixes issue #30).
Ready to receive feedback, I ran your tests, pass, I tested locally if functionality works, works, I can add unit tests if the principle setup is ok for you.

Best regards,
Patrick

### CODE REVIEW

File: APsystemsEZ1/__init__.py [hunk 1]
```diff
@@ -43,6 +43,12 @@ class APsystemsEZ1M:
     power status, alarm information, device information, and power limits.
     """
 
+    @dataclass
+    class DebounceVal:
```
Comment (mawoka-myblock) [inline APsystemsEZ1/__init__.py:47]:
```suggestion
    class _DebounceVal:
```

To make it "private"

File: APsystemsEZ1/__init__.py [hunk 2]
```diff
@@ -65,6 +72,9 @@ def __init__(
         self.session = session
         self.max_power = max_power
         self.min_power = min_power
+        self.debounce_values = debounce_values
+        self._e1 = self.DebounceVal()
+        self._e2 = self.DebounceVal()
```
Comment (mawoka-myblock) [inline APsystemsEZ1/__init__.py:77]:
```suggestion
        self._e1 = self._DebounceVal()
        self._e2 = self._DebounceVal()
```

File: APsystemsEZ1/__init__.py [hunk 3]
```diff
@@ -51,7 +57,8 @@ def __init__(
         max_power: int = 800,
         min_power: int = 30,
         session: ClientSession | None = None,
-    ):
+        debounce_values: bool = False,
```
Comment (mawoka-myblock) [inline APsystemsEZ1/__init__.py:60]:
```suggestion
        enable_debounce: bool = False,
```

File: APsystemsEZ1/__init__.py [hunk 4]
```diff
@@ -65,6 +72,9 @@ def __init__(
         self.session = session
         self.max_power = max_power
         self.min_power = min_power
+        self.debounce_values = debounce_values
```
Comment (mawoka-myblock) [inline APsystemsEZ1/__init__.py:75]:
```suggestion
        self.enable_debounce = enable_debounce
```

File: APsystemsEZ1/__init__.py [hunk 5]
```diff
@@ -101,6 +111,27 @@ async def _request(self, endpoint: str, retry: bool | None = True) -> dict | Non
             if self.session is None:
                 await ses.close()
 
+    def _debounce(self, state: DebounceVal, new_state: float) -> float:
```
Comment (mawoka-myblock) [inline APsystemsEZ1/__init__.py:114]:
```suggestion
    def _debounce(self, state: _DebounceVal, new_state: float) -> float:
```

**Review by mawoka-myblock — CHANGES_REQUESTED** [review]
Awesome, thanks! The cherry on to would be tests! Unit tests of `_debounce()` are/would be enough

**Comment by mawoka-myblock** [comment]
The merge will take some time as I won't merge without tests, but I'll do them eventually if you don't feel like that. But so far, it looks good!

**Comment by patman15** [comment]
Hey!

I'm almost done with tests, just fighting with getting the datetime mocked.
Give me a few hours/days. 
--
Touched, not typed.

**Comment by mawoka-myblock** [comment]
That's soo awesome! Thank you and take your time. We're not in a rush

**Comment by patman15** [comment]
Please check my PR including the unit tests. Happy to receive feedback! Thanks!

**Comment by patman15** [comment]
Just thought again about the PR, will improve the tests a bit further.
--
Touched, not typed.
https://threema.id/ABSJVUAW

**Comment by patman15** [comment]
So, now I'm happy, hope you are satisfied as well!

**Comment by mawoka-myblock** [comment]
Thank you! I'll check the PR soon!

**Comment by mawoka-myblock** [comment]
Oh god, already 3 weeks?! Huge apologies so far! Completely forgot this PR!

**Comment by patman15** [comment]
All good, I'm patient as long as you won't make me redo it a forth(?) time. ;-) Hope you don't do big changes on the main!

**Review by mawoka-myblock — APPROVED** [review]

**Comment by mawoka-myblock** [comment]
Man, I again forgot you! I'm really sorry for the delays here, but y'know, there are other projects. Just some and this right here unfortunately has a low(ish) priority! Sorry for that

**Comment by mawoka-myblock** [comment]
Finally! Thank you soo much for your patience and your work!

**Comment by patman15** [comment]
Thank you for all the work! I will soon bug you with a PR for the HA integration, to enable the behaviour there. :wink:

**Comment by mawoka-myblock** [comment]
> Thank you for all the work! I will soon bug you with a PR for the HA integration, to enable the behavior there. 😉

Just wanted to say that. I'll release a new version of this API now so you can start immediately if you want

**Comment by patman15** [comment]
Currently I can't work on it. I plan to continue at the end of this month.
Do you want me to make the new behaviour, which I think is quite transparent, optional or can I enable it by default in HA? Just to check before I start. :-)

**Comment by mawoka-myblock** [comment]
I think that you can enable it by default

**Pull request merged**